test/functional/sqlsrv/sqlsrv_ae_output_param_sqltype_datetime.phpt

```diff
@@ -0,0 +1,115 @@
+                     "datetime2" => array( "SQLSRV_SQLTYPE_CHAR", "SQLSRV_SQLTYPE_VARCHAR", "SQLSRV_SQLTYPE_NCHAR", "SQLSRV_SQLTYPE_NVARCHAR", "SQLSRV_SQLTYPE_DATETIME", "SQLSRV_SQLTYPE_SMALLDATETIME", "SQLSRV_SQLTYPE_DATE", "SQLSRV_SQLTYPE_TIME", "SQLSRV_SQLTYPE_DATETIMEOFFSET", "SQLSRV_SQLTYPE_DATETIME2" ),
+                     "smalldatetime" => array( "SQLSRV_SQLTYPE_CHAR", "SQLSRV_SQLTYPE_VARCHAR", "SQLSRV_SQLTYPE_NCHAR", "SQLSRV_SQLTYPE_NVARCHAR", "SQLSRV_SQLTYPE_DATETIME", "SQLSRV_SQLTYPE_SMALLDATETIME", "SQLSRV_SQLTYPE_DATE", "SQLSRV_SQLTYPE_TIME", "SQLSRV_SQLTYPE_DATETIMEOFFSET", "SQLSRV_SQLTYPE_DATETIME2" ),
+                     "time" => array( "SQLSRV_SQLTYPE_CHAR", "SQLSRV_SQLTYPE_VARCHAR", "SQLSRV_SQLTYPE_NCHAR", "SQLSRV_SQLTYPE_NVARCHAR", "SQLSRV_SQLTYPE_DATETIME", "SQLSRV_SQLTYPE_SMALLDATETIME", "SQLSRV_SQLTYPE_TIME", "SQLSRV_SQLTYPE_DATETIMEOFFSET", "SQLSRV_SQLTYPE_DATETIME2" ),
+                     "datetimeoffset" => array("SQLSRV_SQLTYPE_CHAR", "SQLSRV_SQLTYPE_VARCHAR", "SQLSRV_SQLTYPE_NCHAR", "SQLSRV_SQLTYPE_NVARCHAR", "SQLSRV_SQLTYPE_DATETIMEOFFSET") );
+
+$conn = AE\connect();
+	
+foreach ($dataTypes as $dataType) {
+    echo "\nTesting $dataType: \n";
+    $success = true;
+
+    // create table
+    $tbname = GetTempTableName("", false);
+    $colMetaArr = array( new AE\ColumnMeta($dataType, "c_det"), new AE\ColumnMeta($dataType, "c_rand", null, false));
+    AE\createTable($conn, $tbname, $colMetaArr);
+	
+    // insert a row
+    $inputValues = array_slice(${explode("(", $dataType)[0] . "_params"}, 1, 2);
+    $r;
+    $stmt = AE\insertRow($conn, $tbname, array( $colMetaArr[0]->colName => $inputValues[0], $colMetaArr[1]->colName => $inputValues[1] ), $r);
+    if ($r === false) {
+        is_incompatible_types_error($dataType, "default type");
+    }
+	
+    // Create a Store Procedure
+    $spname = 'selectAllColumns';
+    $spSql = "CREATE PROCEDURE $spname (@c_det $dataType OUTPUT, @c_rand $dataType OUTPUT ) AS SELECT @c_det = c_det, @c_rand = c_rand FROM $tbname";
+    sqlsrv_query($conn, $spSql);
+
+    // test each SQLSRV_SQLTYPE_ constants
+    foreach ($sqlTypes as $sqlType) {
+        if (!AE\isColEncrypted()) {
+            $isCompatible = false;
+            foreach ($compatList[$dataType] as $compatType) {
+                if (stripos($compatType, $sqlType) !== false) {
+                    $isCompatible = true;
+                }
+            }
+            // 22018 is the SQLSTATE for any incompatible conversion errors
+            if ($isCompatible && sqlsrv_errors()[0]['SQLSTATE'] == 22018) {
+                echo "$sqlType should be compatible with $dataType\n";
+                $success = false;
+            }

```

Comment:
this `if(!AE\isColEncrypted())` block is a bit confusing. The condition `$isCompatible && sqlsrv_errors()[0]['SQLSTATE'] == 22018` is checking if the error exists in the create stored procedure query (since that is the last operation performed and sqlsrv_errors get the error from the last operation), which makes this block of code unnecessary since rarely there will be something wrong with creating a stored procedure. This block of code should be moved down to after the store procedure is called to see if there is any error from calling the store procedure.